### PR TITLE
[eas-cli] add branch:delete command

### PR DESCRIPTION
**Pull request by jkhales**
# Why

Add `eas branch:delete <name>`

# How

delete the branch name.

Added an additional `no/yes` toggle as this is a sensitive operation.

# Test Plan

created and deleted branches.

Confirmed it doesn't delete a branch which a channel is pointing at.
Confirmed it deletes a branch with/without updates on it.

### CODE REVIEW

File: packages/eas-cli/src/commands/branch/delete.ts [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+      }));
+    }
+
+    const { branchId } = await getBranchInfoAsync({ appId: projectId, name });
+
+    if (!jsonFlag) {
+      Log.addNewLineIfNone();
+      Log.warn(
+        `You are about to permamently delete branch: "${name}" and all of the updates published on it.` +
+          `\nThis action is irreversable.`
```
Comment (byCedric) [inline packages/eas-cli/src/commands/branch/delete.ts:156]: I'm not sure if we still have it in Expo CLI, but making this text might help on the emphasis?

**Review by byCedric — APPROVED** [review]

File: packages/eas-cli/src/commands/branch/delete.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+async function getBranchInfoAsync({
+  appId,
+  name,
+}: {
+  appId: string;
+  name: string;
+}): Promise<{ branchId: string }> {
+  const data = await withErrorHandlingAsync(
+    graphqlClient
+      .mutation<
```
Comment (fson) [inline packages/eas-cli/src/commands/branch/delete.ts:23]:
```suggestion
      .query<
```

File: packages/eas-cli/src/commands/branch/delete.ts [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  );
+
+  const {
+    app: {
+      byId: {
+        updateBranchByName: { id: branchId },
+      },
+    },
+  } = data;
+  return { branchId };
```
Comment (fson) [inline packages/eas-cli/src/commands/branch/delete.ts:68]:
```suggestion
  return { branchId: data.app.byId.updateBranchByName.id };
```
Seems less convoluted than destructuring in this case.

File: packages/eas-cli/src/commands/branch/delete.ts [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+                name: string;
+                updates: Pick<Update, 'id'>[];
+              };
+            };
+          };
+        },
+        {
+          appId: string;
+          name: string;
+        }
```
Comment (fson) [inline packages/eas-cli/src/commands/branch/delete.ts:38]:
```suggestion
        ViewBranchQuery,
        ViewBranchQueryVariables
```
`graphql-codegen` will generate these types automatically for you, no need to write and maintain them by hand.

Add this file to the paths of files with GraphQL queries and then re-run `yarn generate-graphql-code`. The config is here:
https://github.com/expo/eas-cli/blob/2e25ed81ca52d9d25afeadd751248d885ed152a9/packages/eas-cli/graphql-codegen.yml#L3-L5

File: packages/eas-cli/src/commands/branch/delete.ts [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+  branchId: string;
+}): Promise<DeleteUpdateBranchResult> {
+  const data = await withErrorHandlingAsync(
+    graphqlClient
+      .mutation<
+        { updateBranch: { deleteUpdateBranch: DeleteUpdateBranchResult } },
+        {
+          branchId: string;
+        }
+      >(
```
Comment (fson) [inline packages/eas-cli/src/commands/branch/delete.ts:83]:
```suggestion
      .mutation<DeleteUpdateBranchMutation, DeleteUpdateBranchMutationVariables>(
```

File: packages/eas-cli/src/commands/branch/delete.ts [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+  static args = [
+    {
+      name: 'name',
+      required: false,
+      description: 'Name of the branch to delete',
+    },
+  ];
+  static flags = {
+    json: flags.boolean({
+      description: `return a json with the edited branch's ID and name.`,
```
Comment (fson) [inline packages/eas-cli/src/commands/branch/delete.ts:115]:
```suggestion
      description: `return JSON with the edited branch's ID and name.`,
```

**Review by fson — CHANGES_REQUESTED** [review]

File: packages/eas-cli/src/commands/branch/delete.ts (hunk 4 above)
Comment (jkhales) [reply packages/eas-cli/src/commands/branch/delete.ts]: Nice!
I added the whole commands sub-directory:
`'src/commands/**/*.ts'`

Will go through and convert some more types in a follow up PR.

File: packages/eas-cli/src/commands/branch/delete.ts (hunk 1 above)
Comment (jkhales) [reply packages/eas-cli/src/commands/branch/delete.ts:132]: That's a fair point, for now will keep like this for consistency as other parts of the CLI output like this.

There is this general issue here as well: https://linear.app/expo/issue/ENG-205/standardize-eas-cli-output

File: packages/eas-cli/src/commands/channel/create.ts [hunk 7]
```diff
@@ -36,7 +36,7 @@ async function createUpdateChannelOnAppAsync({
         { appId: string; name: string; branchMapping: string }
       >(
         gql`
-          mutation CreateUpdateChannelForApp($appId: ID!, $name: String!, $branchMapping: String!) {
+          mutation CreateUpdateChannelOnApp($appId: ID!, $name: String!, $branchMapping: String!) {
```
Comment (jkhales) [inline packages/eas-cli/src/commands/channel/create.ts:39]: fix namespace conflict

**Review by fson — APPROVED** [review]

**Pull request merged**